os/20211014-s390x-secure-execution.md

```diff
@@ -0,0 +1,157 @@
+
+
+##  Build
+ To build SecureExecution Ready image new command is added to `coreos-assembler`:
+ ```
+ cosa buildextend-se --hostkey UNIVERSAL_KEY.crt
+ ```
+ Which is similar to `cosa buildextend-qemu` and also generates qcow2 image.
+
+The qcow2 image comes with `sd-boot` isntalled into its own partition, and LUKS2 pre-encrypted boot and rootfs filesystems.
+Here are partitions of the SecureExecution qcow2 image provided by RH/IBM:
+
+```
+NAME             LABEL        TYPE  MOUNTPOINT
+vdb                           disk
+|-vdb1           se           part
+|-vdb3           crypt_bootfs part
+| `-crypt_bootfs boot         crypt /boot
+`-vdb4           crypt_rootfs part
+  `-crypt_rootfs root         crypt /sysroot
+
+```
+
+- `se`   - This is a new partition `ext4`, where we store a `sd-boot` encrypted image
+- `boot` - boot partition is the same as one on reqular qcow2 image, additionally it's encrypted with random LUKS key during CoreOS build process
+- `root` - rootfs partition is almost the same as one on reqular qcow2 image, additionally it's encrypted with random LUKS key during CoreOS build process
+and contains `/etc/crypttab`, `/etc/luks/root` and `/etc/luks/boot` files.
+
+The reason why `boot` has to be encrypted is that without encryption hacker is able to mount it and modify its contents, for example add new kargs to BLS config file(s), so during system's lifetime that modification could affect `sd-boot` image.
+
+## Firstboot
+
+During guest provisioning new hostkey(s) is/are fetched and used for generation of `sd-boot`. User should specify them within ignition config with pre-defined name template: `/etc/se-hostkeys/ibm-z-hostkey-N` where `N` is a key number, several hostkeys could be used:
+
+```yaml
+storage:
+  files:
+    -
+      path: /etc/se-hostkeys/ibm-z-hostkey-1
+      overwrite: true
+      contents:
+        source: https://url.com/key1
+```
+
+System would also set a new random LUKS keys for both `boot` and `root` filesystems, so each node has its own LUKS keys.
+
+## Lifetime
+
+SecureExecution is enabled by 2 tools provided by IBM's s390-utils package:
+- `genprotimg` - tool to generate `sd-boot`
+- `zipl` - tool to install new bootloader
+
+Despite `zipl` is well known and used by CoreOS, it couldn't be executed same way as before. When system runs in SecureExecution, `ostree` and `rdcore` automatically detect that and perform all required steps.
+That means, that user won't notice any difference, both mentioned components would automatically regenerate `sd-boot` image and
+install it with `zipl` whenever required.
+
+## Ignition Protection
+
+Ignition config will be built and when saved encrypted with a customer's public key.
+The private key is part of the sd-boot image and is generated by RH/IBM during qcow2 image build. The public key is part of the build metadata.
+During firstboot the Ignition config gets decrypted and the key gets removed immediately after it. This could be done with a
```

Comment:
This section is incomplete.  I no longer recall the details; is this still included in the plan?

Reply:
as step4. But this is also to be discussed later.